### DISCUSSION TITLE
Dynamically generate files based on page content

### DISCUSSION THREAD

**Discussion by akhmerov** [mkdocs/mkdocs#2736] [Q&A]
I am experimenting with a plugin that would execute code in the source markdown files and insert back the execution results using jupyter display system. Unlike similar mkdocs-jupyter and mknotebooks plugins I would like to avoid inlining computed assets such as images, and instead save them on disk. Because execution is slow, I would like to avoid reexecuting files that didn't change. Conceptually the plugin should take in the page markdown and produce modified markdown + a bunch of extra files.

This brings me to the main question: what is the proper combination of build events that I should use? So far I hooked into `on_page_read_source`, which then writes the outputs into `Path(config["site_dir"]) / _execute_outputs`. This seems to work OK, however, because I am not modifying the `files` list, the newly generated files are invisible to post-processing. At the very least I see warnings like `Documentation file 'test.md' contains a link to '../_execute_outputs/test.md_1_1.png' which is not found in the documentation files.`

What would be a better way? It seems that I could add the missing files manually at a later stage for example by mutating input in `on_page_markdown`, but this seems really bad. Any other alternatives that I miss?

**Comment by pawamoy** [comment]
Some projects you could use or get inspiration from:

- https://github.com/axiros/docutools
- https://github.com/oprypin/mkdocs-gen-files
- https://github.com/oprypin/mkdocs-literate-nav

**Reply by akhmerov** [reply]
Excellent, thanks! This seems to be quite enough to go on.

**Reply by akhmerov** [reply]
Upon a closer inspection I didn't find a clear answer. To summarize what I found:

### `docutools`

I had a bit of a hard time going through this one, because it seems to apply its own processing that doesn't follow mkdocs conventions and seems to do things in places where mkdocs doesn't expect this (e.g. reading sources in `on_config`). I'd like to find a solution that plays nicely with other plugins, and therefore this doesn't seem to be a way to go.

The only interaction with `files` that it uses within the plugin system, is to [hide the auto-generated files](https://github.com/axiros/docutools/blob/f170fcc203b6b213ea611c8bc1cbee65bf23db78/src/lcdoc/mkdocs/lp/__init__.py#L721-L724) from mkdocs. These auto-generated files are placed in the source directory, something that I also want to avoid.

### `mkdocs-gen-files`

Performs execution on `on_files` event inside a temporary directory ([source](https://github.com/oprypin/mkdocs-gen-files/blob/master/mkdocs_gen_files/plugin.py#L26-L38)). This has a downside of not being able to detect changes since all files are run unconditionally. The idea of keeping track of generated files separately seems to be useful, but doesn't answer the question of how to interact with the event system.

### `mkdocs-literate-nav`

Doesn't really generate new files, but rather [checks](https://github.com/oprypin/mkdocs-literate-nav/blob/master/mkdocs_literate_nav/plugin.py#L48) if any other plugin did it in a way that this one isn't aware about.

**Reply by pawamoy** [reply]
Sorry you didn't find what you were looking for. Don't hesitate to unmark my reply as the answer, if that's possible. You can also try to ask for help on the gitter channel, at https://gitter.im/mkdocs/community.

**Reply by akhmerov** [reply]
Upon further examining the source, I decided to mutate `files` in `on_page_markdown`.
